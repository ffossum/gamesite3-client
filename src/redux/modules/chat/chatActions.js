/* @flow */
export const JOIN_CHANNEL = 'chat/join channel';
export const LEAVE_CHANNEL = 'chat/leave channel';
export const SEND_MESSAGE = 'chat/send message';
export const SEND_GAME_MESSAGE = 'chat/send game message';
export const RECEIVE_MESSAGE = 'chat/receive message';
export const CLEAR_CHAT = 'chat/clear';

export type ChannelName = string;
export type UserMessage = {
  ch: ChannelName,
  txt: string,
  uid: string,
};

export type JoinChannelAction = {
  type: typeof JOIN_CHANNEL,
  payload: string,
};
export function joinChannel(channelName: ChannelName): JoinChannelAction {
  return {
    type: JOIN_CHANNEL,
    payload: channelName,
  };
}

export type LeaveChannelAction = {
  type: typeof LEAVE_CHANNEL,
  payload: string,
};
export function leaveChannel(channelName: ChannelName): LeaveChannelAction {
  return {
    type: LEAVE_CHANNEL,
    payload: channelName,
  };
}

export type SendMessageAction = {
  type: typeof SEND_MESSAGE,
  payload: UserMessage,
};
export function sendMessage(
  userId: string,
  channelName: string,
  messageText: string,
): SendMessageAction {
  return {
    type: SEND_MESSAGE,
    payload: {
      ch: channelName,
      txt: messageText,
      uid: userId,
    },
  };
}

export type SendGameMessageAction = {
  type: typeof SEND_GAME_MESSAGE,
  payload: {
    players: string[],
    gameId: string,
    userId: string,
    text: string,
  },
};
export function sendGameMessage(
  userId: string,
  gameId: string,
  players: string[],
  text: string,
) {
  return {
    type: SEND_GAME_MESSAGE,
    payload: {
      userId,
      gameId,
      players,
      text,
    },
  };
}
export type ReceiveMessageAction = {
  type: typeof RECEIVE_MESSAGE,
  payload: {
    msg: UserMessage,
    time: string,
  },
};
export function receiveMessage(
  msg: UserMessage,
  time: string,
): ReceiveMessageAction {
  return {
    type: RECEIVE_MESSAGE,
    payload: {
      msg,
      time,
    },
  };
}

export type ClearChatAction = {
  type: typeof CLEAR_CHAT,
  payload: string,
};
export function clearChat(channelName: string) {
  return {
    type: CLEAR_CHAT,
    payload: channelName,
  };
}

export type ChatAction =
  | JoinChannelAction
  | LeaveChannelAction
  | SendMessageAction
  | SendGameMessageAction
  | ReceiveMessageAction
  | ClearChatAction;
