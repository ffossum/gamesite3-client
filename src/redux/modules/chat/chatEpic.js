/* @flow */
import { combineEpics } from 'redux-observable';
import type { Observable } from 'rxjs';
import type { Store } from 'redux';

import { SEND_MESSAGE, JOIN_CHANNEL, receiveMessage } from './chatActions';
import type { SendMessageAction, JoinChannelAction } from './chatActions';
import type DeepstreamClient from '../../deepstreamClient';

type Dependencies = {
  deepstreamClient: DeepstreamClient,
};

function sendMessageEpic(
  action$: Observable<*>,
  store: Store<*, *>,
  { deepstreamClient }: Dependencies,
) {
  return action$
    .filter(action => action.type === SEND_MESSAGE)
    .do((action: SendMessageAction) => {
      deepstreamClient.emit(`chat:${action.payload.ch}`, {
        t: 'chatmsg',
        ...action.payload,
      });
    })
    .ignoreElements();
}

export function joinChannelEpic(
  action$: Observable<*>,
  store: Store<*, *>,
  { deepstreamClient }: Dependencies,
) {
  return action$
    .filter(action => action.type === JOIN_CHANNEL)
    .flatMap((action: JoinChannelAction) => {
      const channelName = `chat:${action.payload}`;
      return deepstreamClient.subscribe(channelName);
    })
    .flatMap(data => {
      switch (data.t) {
        case 'chatmsg': {
          const time = new Date().toISOString();
          return [receiveMessage(data, time)];
        }

        default:
          return [];
      }
    });
}

export default combineEpics(sendMessageEpic, joinChannelEpic);
