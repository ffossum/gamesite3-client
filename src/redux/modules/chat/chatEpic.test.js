/* @flow */
/* eslint-env jest */
import { Observable } from 'rxjs';
import { ActionsObservable } from 'redux-observable';
import chatEpic from './chatEpic';
import { sendMessage, joinChannel, receiveMessage } from './chatActions';

describe('chat epic', () => {
  let deepstreamClient;
  const store = null;

  beforeEach(() => {
    deepstreamClient = {
      emit: jest.fn(),
      subscribe: jest.fn(),
    };
  });

  test('emits send message event to deepstream', async () => {
    const action = sendMessage('asdf-id', 'general', 'message text');
    const action$ = ActionsObservable.of(action);

    await new Promise(resolve => {
      chatEpic(action$, store, { deepstreamClient })
        .toArray()
        .subscribe(actions => {
          expect(deepstreamClient.emit).toHaveBeenCalledTimes(1);
          expect(deepstreamClient.emit.mock.calls).toMatchSnapshot();
          expect(actions).toHaveLength(0);
          resolve();
        });
    });
  });

  test('subscribes to chat channel deepstream events', async () => {
    const action = joinChannel('general');
    const action$ = ActionsObservable.of(action);

    deepstreamClient.subscribe.mockReturnValueOnce(Observable.empty());

    await new Promise(resolve => {
      chatEpic(action$, store, { deepstreamClient })
        .toArray()
        .subscribe(actions => {
          expect(deepstreamClient.subscribe).toHaveBeenCalledTimes(1);
          expect(actions).toHaveLength(0);

          resolve();
        });
    });
  });

  test('handles events from deepstream subscription', async () => {
    const action = joinChannel('general');
    const action$ = ActionsObservable.of(action);

    const eventData = {
      t: 'chatmsg',
      ch: 'general',
      uid: 'asdf-id',
      txt: 'message text',
    };
    deepstreamClient.subscribe.mockReturnValueOnce(Observable.of(eventData));

    await new Promise(resolve => {
      chatEpic(action$, store, { deepstreamClient })
        .toArray()
        .subscribe(actions => {
          expect(deepstreamClient.subscribe).toHaveBeenCalledTimes(1);

          expect(actions).toHaveLength(1);
          expect(actions[0]).toEqual(receiveMessage(eventData));

          resolve();
        });
    });
  });
});
