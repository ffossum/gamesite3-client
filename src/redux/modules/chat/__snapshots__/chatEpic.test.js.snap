// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat epic emits send message event to deepstream 1`] = `
Array [
  Array [
    "chat:general",
    Object {
      "ch": "general",
      "t": "chatmsg",
      "txt": "message text",
      "uid": "asdf-id",
    },
  ],
]
`;
